src/praisonai-agents/praisonaiagents/tools/registry.py

```
@@ -245,6 +256,18 @@ def get_tool_definitions(self) -> List[Dict[str, Any]]:
                         continue
                 
+                # Deny = hide: drop tools the permission resolver rejects so the
+                # model is never offered a tool it cannot call. Same filtering
+                # seam as check_availability() above. Fail-open on resolver error
+                # (keep the tool; call-time gate still enforces).
+                if permission_resolver is not None:
+                    try:
+                        if not permission_resolver(entry.name):
+                            logging.debug(f"Tool '{entry.name}' hidden by permission policy")
+                            continue
+                    except Exception as e:
```

Ruff (0.15.20)

[warning] Do not catch blind exception: `Exception`

(BLE001)